shimmingtoolbox/coils/spherical_harmonics.py

```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/python3
+# -*- coding: utf-8 -*
+
+import numpy as np
+import math
+from scipy.special import factorial2
+from scipy.special import lpmv
+
+
+def spherical_harmonics(orders, x, y, z):
+    """
+    Returns an array of spherical harmonic basis fields with the order/degree index along the 4th dimension.
+
+    Args:
+        orders (numpy.ndarray):  Degrees of the desired terms in the series expansion, specified as a vector of
+                                 non-negative integers (`np.array(range(0, 3))` yields harmonics up to (n-1)-th order).
+                                 Must be non negative.
+        x (numpy.ndarray): 3-D arrays of grid coordinates
+        y (numpy.ndarray): 3-D arrays of grid coordinates (same shape as x)
+        z (numpy.ndarray): 3-D arrays of grid coordinates (same shape as x)
+
+    Returns:
+        numpy.ndarray: 4d basis set of spherical harmonics with order/degree ordered along 4th dimension
+
+    Examples:
+        Initialize grid positions
+
+        >>> [x, y, z] = np.meshgrid(np.array(range(-10, 11)), np.array(range(-10, 11)), np.array(range(-10, 11)), indexing='ij')
+
+        0th-to-2nd order terms inclusive
+
+        >>> orders = np.array(range(0, 3))
+        >>> basis = spherical_harmonics(orders, x, y, z)
+
+    Notes:
+        - basis[:, :, :,0] corresponds to the 0th-order constant term (globally=unity)
+            - 0: *c*
+
+        - basis[:, :, :, 1:3] to 1st-order linear terms
+            - 1: *y*
+            - 2: *z*
+            - 3: *x*
+
+        - basis[:, :, :, 4:8] to 2nd-order terms
+            - 4: *xy*
+            - 5: *zy*
+            - 6: *z2*
+            - 7: *zx*
+            - 8: *x2y2*
+
+        Based on
+            - spherical_harmonics.m by topfer@ualberta.ca
+            - calc_spherical_harmonics_arb_points_cz.m by jaystock@nmr.mgh.harvard.edu
+    """
+
+    def number_of_coef(orders):
+        out = 0
+        for n in orders:
+            m = 2 * n + 1
+            out += m
+
+        return out
```

Comment:
I'd recommend a clearer variable name than `out`, reflecting what the output is:
```suggestion
        n_coef = 0
        for n in orders:
            m = 2 * n + 1
            n_coef += m

        return n_coef
```